From 968018831ea1dc22dbee896b7faf7d5bb4b2530d Mon Sep 17 00:00:00 2001
From: "K.Takata" <kentkt@csc.jp>
Date: Sun, 4 Nov 2018 20:24:52 +0900
Subject: [PATCH 35/35] nsis,dosinst: Make uninstaller silent when installing

* Write "AllowSilent=1" to the registry so that it describe that this
  uninstaller supports silent mode.
* If the installer found "AllowSilent=1", kick the uninstaller with
  the silent mode.
* Delete only the vimfiles directory in the installed directory.
  Not in the $HOME directory.
* Adjust some English messages.
---
 nsis/gvim.nsi             |  15 ++---
 nsis/lang/danish.nsi      |   3 +-
 nsis/lang/dutch.nsi       |   6 +-
 nsis/lang/english.nsi     |   6 +-
 nsis/lang/german.nsi      |   4 +-
 nsis/lang/italian.nsi     |   6 +-
 nsis/lang/japanese.nsi    |   4 +-
 nsis/lang/simpchinese.nsi |   6 +-
 nsis/lang/tradchinese.nsi |   6 +-
 src/dosinst.c             | 134 ++++++++++++++++++++++++++++----------
 10 files changed, 125 insertions(+), 65 deletions(-)

diff --git a/nsis/gvim.nsi b/nsis/gvim.nsi
index f1a79682a..a64f2537c 100644
--- a/nsis/gvim.nsi
+++ b/nsis/gvim.nsi
@@ -290,7 +290,7 @@ Section "$(str_section_old_ver)" id_section_old_ver
 	# run the install program to check for already installed versions
 	SetOutPath $TEMP
 	File /oname=install.exe ${VIMSRC}\installw32.exe
-	DetailPrint "$(str_msg_wait_uninst)"
+	DetailPrint "$(str_msg_uninstalling)"
 	${Do}
 	  nsExec::Exec "$TEMP\install.exe -uninstall-check"
 	  Pop $3
@@ -649,12 +649,13 @@ Section -post
 	${EndIf}
 !endif
 
-	# Register EstimatedSize.
+	# Register EstimatedSize and AllowSilent.
 	# Other information will be set by the install.exe (dosinst.c).
 	${If} ${RunningX64}
 	  SetRegView 64
 	${EndIf}
 	WriteRegDWORD HKLM "${UNINST_REG_KEY_VIM}" "EstimatedSize" $3
+	WriteRegDWORD HKLM "${UNINST_REG_KEY_VIM}" "AllowSilent" 1
 	${If} ${RunningX64}
 	  SetRegView lastused
 	${EndIf}
@@ -1080,14 +1081,8 @@ Section "un.$(str_unsection_vimfiles)" id_unsection_vimfiles
 	Call un.GetParent
 	Pop $1
 
-	# if a plugin dir was created at installation ask the user to remove it
-	# first look in the root of the installation then in HOME
-	${IfNot} ${FileExists} $1\vimfiles
-	  ReadEnvStr $1 "HOME"
-	${EndIf}
-
-	${If} $1 != ""
-	${AndIf} ${FileExists} $1\vimfiles
+	# if a plugin dir was created at installation remove it
+	${If} ${FileExists} $1\vimfiles
 	  RMDir $1\vimfiles\colors
 	  RMDir $1\vimfiles\compiler
 	  RMDir $1\vimfiles\doc
diff --git a/nsis/lang/danish.nsi b/nsis/lang/danish.nsi
index 7b218fa2d..7669fc98d 100644
--- a/nsis/lang/danish.nsi
+++ b/nsis/lang/danish.nsi
@@ -148,7 +148,8 @@ LangString str_msg_rm_exe_fail   ${LANG_DANISH}     "Nogle filer i $0 er ikke bl
 
 #LangString str_msg_rm_root_fail  ${LANG_DANISH}     "ADVARSEL: Kan ikke fjerne $\"$vim_install_root$\", den er ikke tom!"
 
-LangString str_msg_wait_uninst   ${LANG_DANISH}     "Venter på at den gamle version bliver afinstalleret..."
+LangString str_msg_uninstalling  ${LANG_DANISH} \
+    "Uninstalling the old version..."
 
 LangString str_msg_registering   ${LANG_DANISH}     "Registrerer..."
 
diff --git a/nsis/lang/dutch.nsi b/nsis/lang/dutch.nsi
index f37387eb4..1f3ffbfa0 100644
--- a/nsis/lang/dutch.nsi
+++ b/nsis/lang/dutch.nsi
@@ -149,7 +149,7 @@ LangString str_desc_rm_exe          ${LANG_DUTCH} \
 LangString str_unsection_vimfiles   ${LANG_DUTCH} \
     "Remove vimfiles directory"
 LangString str_desc_rm_vimfiles     ${LANG_DUTCH} \
-    "Remove your vimfiles directory if it is empty."
+    "Remove the vimfiles directory if it is empty."
 
 LangString str_unsection_rootdir    ${LANG_DUTCH} \
     "Remove the Vim root directory"
@@ -221,8 +221,8 @@ LangString str_msg_rm_exe_fail   ${LANG_DUTCH} \
 #LangString str_msg_rm_root_fail  ${LANG_DUTCH} \
 #    "WAARSCHUWING: Kan $\"$vim_install_root$\" niet verwijderen omdat het niet leeg is!"
 
-LangString str_msg_wait_uninst   ${LANG_DUTCH} \
-    "Waiting the old version to be uninstalled..."
+LangString str_msg_uninstalling  ${LANG_DUTCH} \
+    "Uninstalling the old version..."
 
 LangString str_msg_registering   ${LANG_DUTCH} \
     "Registering..."
diff --git a/nsis/lang/english.nsi b/nsis/lang/english.nsi
index 88fed67a4..859b28d7a 100644
--- a/nsis/lang/english.nsi
+++ b/nsis/lang/english.nsi
@@ -157,7 +157,7 @@ LangString str_desc_rm_exe          ${LANG_ENGLISH} \
 LangString str_unsection_vimfiles   ${LANG_ENGLISH} \
     "Remove vimfiles directory"
 LangString str_desc_rm_vimfiles     ${LANG_ENGLISH} \
-    "Remove your vimfiles directory if it is empty."
+    "Remove the vimfiles directory if it is empty."
 
 LangString str_unsection_rootdir    ${LANG_ENGLISH} \
     "Remove the Vim root directory"
@@ -230,8 +230,8 @@ LangString str_msg_rm_exe_fail   ${LANG_ENGLISH} \
 #LangString str_msg_rm_root_fail  ${LANG_ENGLISH} \
 #    "WARNING: Cannot remove $\"$vim_install_root$\", it is not empty!"
 
-LangString str_msg_wait_uninst   ${LANG_ENGLISH} \
-    "Waiting the old version to be uninstalled..."
+LangString str_msg_uninstalling  ${LANG_ENGLISH} \
+    "Uninstalling the old version..."
 
 LangString str_msg_registering   ${LANG_ENGLISH} \
     "Registering..."
diff --git a/nsis/lang/german.nsi b/nsis/lang/german.nsi
index 02079d593..b188fab33 100644
--- a/nsis/lang/german.nsi
+++ b/nsis/lang/german.nsi
@@ -220,8 +220,8 @@ LangString str_msg_rm_exe_fail   ${LANG_GERMAN} \
 #    "Achtung: Kann Verzeichnis $\"$vim_install_root$\" nicht entfernen, \
 #     weil es nicht leer ist!"
 
-LangString str_msg_wait_uninst   ${LANG_GERMAN} \
-    "Warte, dass alte Version deinstalliert wird..."
+LangString str_msg_uninstalling  ${LANG_GERMAN} \
+    "Uninstalling the old version..."
 
 LangString str_msg_registering   ${LANG_GERMAN} \
     "Registriere..."
diff --git a/nsis/lang/italian.nsi b/nsis/lang/italian.nsi
index 28dfb280b..5cbdba056 100644
--- a/nsis/lang/italian.nsi
+++ b/nsis/lang/italian.nsi
@@ -156,7 +156,7 @@ LangString str_desc_rm_rc           ${LANG_ITALIAN} \
 LangString str_unsection_vimfiles   ${LANG_ITALIAN} \
     "Remove vimfiles directory"
 LangString str_desc_rm_vimfiles     ${LANG_ITALIAN} \
-    "Remove your vimfiles directory if it is empty."
+    "Remove the vimfiles directory if it is empty."
 
 LangString str_unsection_rootdir    ${LANG_ITALIAN} \
     "Remove the Vim root directory"
@@ -230,8 +230,8 @@ LangString str_msg_rm_exe_fail   ${LANG_ITALIAN} \
 #LangString str_msg_rm_root_fail  ${LANG_ITALIAN} \
 #    "AVVISO: Non posso cancellare $\"$vim_install_root$\", non è vuota!"
 
-LangString str_msg_wait_uninst   ${LANG_ITALIAN} \
-    "Waiting the old version to be uninstalled..."
+LangString str_msg_uninstalling  ${LANG_ITALIAN} \
+    "Uninstalling the old version..."
 
 LangString str_msg_registering   ${LANG_ITALIAN} \
     "Registering..."
diff --git a/nsis/lang/japanese.nsi b/nsis/lang/japanese.nsi
index ce3725443..ce0b15b9c 100644
--- a/nsis/lang/japanese.nsi
+++ b/nsis/lang/japanese.nsi
@@ -226,8 +226,8 @@ LangString str_msg_rm_exe_fail   ${LANG_JAPANESE} \
 #LangString str_msg_rm_root_fail  ${LANG_JAPANESE} \
 #    "WARNING: Cannot remove $\"$vim_install_root$\", it is not empty!"
 
-LangString str_msg_wait_uninst   ${LANG_JAPANESE} \
-    "古いバージョンがアンインストールされるのを待っています..."
+LangString str_msg_uninstalling  ${LANG_JAPANESE} \
+    "古いバージョンをアンインストールしています..."
 
 LangString str_msg_registering   ${LANG_JAPANESE} \
     "登録中..."
diff --git a/nsis/lang/simpchinese.nsi b/nsis/lang/simpchinese.nsi
index 2f6378603..b8233ea42 100644
--- a/nsis/lang/simpchinese.nsi
+++ b/nsis/lang/simpchinese.nsi
@@ -146,7 +146,7 @@ LangString str_desc_rm_exe          ${LANG_SIMPCHINESE} \
 LangString str_unsection_vimfiles   ${LANG_SIMPCHINESE} \
     "Remove vimfiles directory"
 LangString str_desc_rm_vimfiles     ${LANG_SIMPCHINESE} \
-    "Remove your vimfiles directory if it is empty."
+    "Remove the vimfiles directory if it is empty."
 
 LangString str_unsection_rootdir    ${LANG_SIMPCHINESE} \
     "Remove the Vim root directory"
@@ -217,8 +217,8 @@ LangString str_msg_rm_exe_fail   ${LANG_SIMPCHINESE} \
 #    "警告：无法删除 Vim 安装目录“$vim_install_root”，\
 #     该目录下仍有其他文件。"
 
-LangString str_msg_wait_uninst   ${LANG_SIMPCHINESE} \
-    "Waiting the old version to be uninstalled..."
+LangString str_msg_uninstalling  ${LANG_SIMPCHINESE} \
+    "Uninstalling the old version..."
 
 LangString str_msg_registering   ${LANG_SIMPCHINESE} \
     "Registering..."
diff --git a/nsis/lang/tradchinese.nsi b/nsis/lang/tradchinese.nsi
index 50e0c4cc3..76926a0b0 100644
--- a/nsis/lang/tradchinese.nsi
+++ b/nsis/lang/tradchinese.nsi
@@ -147,7 +147,7 @@ LangString str_desc_rm_exe          ${LANG_TRADCHINESE} \
 LangString str_unsection_vimfiles   ${LANG_TRADCHINESE} \
     "Remove vimfiles directory"
 LangString str_desc_rm_vimfiles     ${LANG_TRADCHINESE} \
-    "Remove your vimfiles directory if it is empty."
+    "Remove the vimfiles directory if it is empty."
 
 LangString str_unsection_rootdir    ${LANG_TRADCHINESE} \
     "Remove the Vim root directory"
@@ -218,8 +218,8 @@ LangString str_msg_rm_exe_fail   ${LANG_TRADCHINESE} \
 #    "警告：無法刪除 Vim 安裝資料夾「$vim_install_root」，\
 #     該資料夾下仍有其他檔案。"
 
-LangString str_msg_wait_uninst   ${LANG_TRADCHINESE} \
-    "Waiting the old version to be uninstalled..."
+LangString str_msg_uninstalling  ${LANG_TRADCHINESE} \
+    "Uninstalling the old version..."
 
 LangString str_msg_registering   ${LANG_TRADCHINESE} \
     "Registering..."
diff --git a/src/dosinst.c b/src/dosinst.c
index 6d5c95ba5..b09a3707b 100644
--- a/src/dosinst.c
+++ b/src/dosinst.c
@@ -450,6 +450,50 @@ window_cb(HWND hwnd, LPARAM lparam)
     return TRUE;
 }
 
+/*
+ * Run the uninstaller silently.
+ */
+    static int
+run_silent_uninstall(char *uninst_exe)
+{
+    char    vimrt_dir[BUFSIZE];
+    char    temp_uninst[BUFSIZE];
+    char    temp_dir[MAX_PATH];
+    char    buf[BUFSIZE * 2 + 10];
+    int	    i;
+    DWORD   tick;
+
+    strcpy(vimrt_dir, uninst_exe);
+    remove_tail(vimrt_dir);
+
+    if (!GetTempPath(sizeof(temp_dir), temp_dir))
+	return FAIL;
+
+    /* Copy the uninstaller to a temporary exe. */
+    tick = GetTickCount();
+    for (i = 0; ; i++)
+    {
+	sprintf(temp_uninst, "%s\\vimun%04X.exe", temp_dir,
+							(i + tick) & 0xFFFF);
+	if (CopyFile(uninst_exe, temp_uninst, TRUE))
+	    break;
+	if (GetLastError() != ERROR_FILE_EXISTS)
+	    return FAIL;
+	if (i == 65535)
+	    return FAIL;
+    }
+
+    /* Run the copied uninstaller silently. */
+    if (strchr(temp_uninst, ' ') != NULL)
+	sprintf(buf, "\"%s\" /S _?=%s", temp_uninst, vimrt_dir);
+    else
+	sprintf(buf, "%s /S _?=%s", temp_uninst, vimrt_dir);
+    run_command(buf);
+
+    DeleteFile(temp_uninst);
+    return OK;
+}
+
 /*
  * Check for already installed Vims.
  * Return non-zero when found one.
@@ -470,6 +514,7 @@ uninstall_check(int skip_question)
     DWORD	value_type;
     DWORD	orig_num_keys;
     DWORD	new_num_keys;
+    DWORD	allow_silent;
     int		foundone = 0;
 
     code = RegOpenKeyEx(HKEY_LOCAL_MACHINE, uninstall_key, 0,
@@ -496,6 +541,16 @@ uninstall_check(int skip_question)
 	    local_bufsize = BUFSIZE;
 	    CHECK_REG_ERROR(code);
 
+	    allow_silent = 0;
+	    if (skip_question)
+	    {
+		DWORD varsize = sizeof(DWORD);
+
+		RegQueryValueEx(uninstall_key_handle, "AllowSilent", 0,
+			&value_type, (LPBYTE)&allow_silent,
+			&varsize);
+	    }
+
 	    foundone = 1;
 	    printf("\n*********************************************************\n");
 	    printf("Vim Install found what looks like an existing Vim version.\n");
@@ -550,45 +605,54 @@ uninstall_check(int skip_question)
 			/* Find existing .bat files before deleting them. */
 			find_bat_exe(TRUE);
 
-			/* Execute the uninstall program.  Put it in double
-			 * quotes if there is an embedded space. */
+			if (allow_silent)
 			{
-			    char buf[BUFSIZE];
-
-			    if (strchr(temp_string_buffer, ' ') != NULL)
-				sprintf(buf, "\"%s\"", temp_string_buffer);
-			    else
-				strcpy(buf, temp_string_buffer);
-			    run_command(buf);
-			}
-
-			/* Count the number of windows with a title that match
-			 * the installer, so that we can check when it's done.
-			 * The uninstaller copies itself, executes the copy
-			 * and exits, thus we can't wait for the process to
-			 * finish. */
-			sleep(1);  /* wait for uninstaller to start up */
-			num_windows = 0;
-			EnumWindows(window_cb, 0);
-			if (num_windows == 0)
-			{
-			    /* Did not find the uninstaller, ask user to press
-			     * Enter when done. Just in case. */
-			    printf("Press Enter when the uninstaller is finished\n");
-			    rewind(stdin);
-			    (void)getchar();
+			    if (run_silent_uninstall(temp_string_buffer)
+								    == FAIL)
+				allow_silent = 0; /* Retry with non silent. */
 			}
-			else
+			if (!allow_silent)
 			{
-			    printf("Waiting for the uninstaller to finish (press CTRL-C to abort).");
-			    do
+			    /* Execute the uninstall program.  Put it in double
+			     * quotes if there is an embedded space. */
+			    {
+				char buf[BUFSIZE];
+
+				if (strchr(temp_string_buffer, ' ') != NULL)
+				    sprintf(buf, "\"%s\"", temp_string_buffer);
+				else
+				    strcpy(buf, temp_string_buffer);
+				run_command(buf);
+			    }
+
+			    /* Count the number of windows with a title that match
+			     * the installer, so that we can check when it's done.
+			     * The uninstaller copies itself, executes the copy
+			     * and exits, thus we can't wait for the process to
+			     * finish. */
+			    sleep(1);  /* wait for uninstaller to start up */
+			    num_windows = 0;
+			    EnumWindows(window_cb, 0);
+			    if (num_windows == 0)
+			    {
+				/* Did not find the uninstaller, ask user to press
+				 * Enter when done. Just in case. */
+				printf("Press Enter when the uninstaller is finished\n");
+				rewind(stdin);
+				(void)getchar();
+			    }
+			    else
 			    {
-				printf(".");
-				fflush(stdout);
-				sleep(1);  /* wait for the uninstaller to finish */
-				num_windows = 0;
-				EnumWindows(window_cb, 0);
-			    } while (num_windows > 0);
+				printf("Waiting for the uninstaller to finish (press CTRL-C to abort).");
+				do
+				{
+				    printf(".");
+				    fflush(stdout);
+				    sleep(1);  /* wait for the uninstaller to finish */
+				    num_windows = 0;
+				    EnumWindows(window_cb, 0);
+				} while (num_windows > 0);
+			    }
 			}
 			printf("\nDone!\n");
 
-- 
2.17.0

